### PR TITLE
Add shared buffer for granular stats and update time measurements.

### PR DESCRIPTION
**Pull request by yzhang1991**

### CODE REVIEW

File: src/ee/execution/VoltDBEngine.h [hunk 1]
```diff
@@ -568,12 +575,20 @@ class __attribute__((visibility("default"))) VoltDBEngine {
         /** buffer object for exceptions generated by the EE **/
         ReferenceSerializeOutput m_exceptionOutput;
 
+        ReferenceSerializeOutput m_granularStatsOutput;
```
Comment (yzhang1991) [inline src/ee/execution/VoltDBEngine.h]: Need a comment here

File: src/ee/execution/VoltDBEngine.cpp [hunk 2]
```diff
@@ -395,11 +400,17 @@ int VoltDBEngine::executePlanFragments(int32_t numFragments,
...
+        startTime = std::chrono::high_resolution_clock::now();
         if (executePlanFragment(planfragmentIds[m_currentIndexInBatch],
                                 inputDependencyIds ? inputDependencyIds[m_currentIndexInBatch] : -1,
                                 m_currentIndexInBatch == 0,
                                 m_currentIndexInBatch == (numFragments - 1))) {
             ++failures;
+        }
+        endTime = std::chrono::high_resolution_clock::now();
+        elapsedNanoseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime);
+        m_granularStatsOutput.writeLong(elapsedNanoseconds.count());
```
Comment (yzhang1991) [inline src/ee/execution/VoltDBEngine.cpp]: Surprisingly, we use nanoseconds in procedure statistics. I used microseconds at first too.

File: tests/ee/test_utils/plan_testing_baseclass.h [hunk 3]
```diff
@@ -129,12 +129,15 @@ class PlanTestingBaseClass : public Test {
         m_topend.reset(TOPEND::newInstance());
         m_engine.reset(new voltdb::VoltDBEngine(m_topend.get()));
         m_parameter_buffer.reset(new char [4 * 1024]);
+        m_granular_stats_buffer.reset(new char [4 * 1024]);
```
Comment (yzhang1991) [inline tests/ee/test_utils/plan_testing_baseclass.h]: In the C++ test pull request this is already replaced by a variable.

File: src/ee/execution/VoltDBEngine.h [hunk 4]
```diff
@@ -259,9 +259,16 @@ class __attribute__((visibility("default"))) VoltDBEngine {
                 voltdb::VOLT_EE_EXCEPTION_TYPE_NONE;
         }
 
+        void resetGranularStatisticsOutputBuffer() {
```
Comment (yzhang1991) [inline src/ee/execution/VoltDBEngine.h]: perBatch

File: src/ee/execution/VoltDBEngine.cpp (hunk 2 above)
Comment (wolffcm) [inline src/ee/execution/VoltDBEngine.cpp]: Oh, now I know this.

**Comment by wolffcm** [comment]
I forgot to mention that it would be good to touch base with Phil before merging since this could have a performance impact.

**Comment by wolffcm** [comment]
I thought of another comment that I forgot to make.  Is the layout of the perFragmentStats buffer described in a comment anywhere?  Seems like it would be good to state that explicitly somewhere.

**Comment by yzhang1991** [comment]
I will ask Phil to enable the perf test on Jenkins.

**Comment by yzhang1991** [comment]
Merged with command line tools.

**Pull request closed without merging**